### ISSUE TITLE
Make common file_base smarter

### ISSUE THREAD

**Issue by aeslaughter** [idaholab/moose#4215]
If file_base is specified at the top-level in the [Outputs] block the block name should still be append.

``` text
[Outputs]
  file_base = my_filename
  [./displaced] # this should create a file: my_filename_displaced.e
    type = Exodus
    use_displaced = true
  [../]
[]
```

**Comment by permcody** [comment]
Hmm - is this really what you want @aeslaughter? This ticket might need to be closed.

**Comment by friedmud** [comment]
I actually kind of like this idea.

**Comment by aeslaughter** [comment]
It was an idea I had a long time ago that I think would be nice. Currently, if file_base is set at the top level then all file output use the same name, which creates collisions easily.

**Comment by GiudGiud** [comment]
@idaholab/moose-ccb we're considering doing that. 
Any opposition? 

This makes Outputs/file_base differ from Outputs/my_object_name/file_base, which is probably the only downside to me. 
Maybe we should rename Output/file_base to Ouputs/file_prefix

**Comment by cticenhour** [comment]
The two-level `file_base` confusion needs to be addressed if we do this, like you noted, but I would be generally supportive of this. 

Regarding naming I see two paths (amongst many others, probably, but these two occured to me first):

- `file_prefix` / `file_base` (@GiudGiud's proposal): This seems distinct enough that it would be understood by the majority of our users. "Base" has pretty consistently been used up to this point in cases where nothing more is getting appended or modified to it, so it being the sub-level parameter "end point" is reasonable.
- `file_base_prefix` / `file_base`: As I could still see some people being confused on `file_prefix` (top level) vs `file_base` (sub-level), this option seems more self-documenting to me, but less concise.

There's also an edge case here that we might want to note. What happens when we have the situation:

```
[Outputs]
  file_base = out
  exodus = true
[]
```

or similar for other shorthand syntax? Currently, this would simply output `out.e` (overriding the `exodus = true` convention of `input_out.e`), so we'll need to decide how we want to handle it with a change in top-level parameter behavior. I doubt `out_input_out.e` is how we want to do it - perhaps `out_exodus.e`?

**Comment by GiudGiud** [comment]
`file_base_prefix` or `file_prefix` both work for me. They'd be a parameter rename. 

@kyriv-lab thoughts on the edge case? 
Maybe we should preserve the existing behavior here (out.e). `exodus` is not really the name of the object in the input
So we'd need to detect: 
- object is created by shortcut action
- file_base is passed in prefix

**Comment by joshuahansel** [comment]
I'm a bit confused some of this discussion, but here's my preference:
```
[Outputs]
  file_base = blah
  exodus = true # produces <parent_file_base>.e; let "default" <parent_file_base> be <input> (or <input>_out if needed)
  [ex]
    type = Exodus
    file_base = foo # produces <child_file_base>.e or if not provided, <parent_file_base>_<block_name>.e
  []
[]
```
I haven't made any considerations of backwards compatibility here - just the preference in a clean slate.

Also, if we don't already do this, keep a tab of output file names and check for no collisions.

**Comment by GiudGiud** [comment]
see the PR for the rules.
This is still opt-in so it's not "smarter" on its own

**Comment by kyriv-lab** [comment]
file_base resolution rules for [Outputs] (#4215)

Problem: a common file_base set at the top of [Outputs] was inherited verbatim by every sub-block that didn't set its own, so two outputs could silently collide on the same filename and fail with an unhelpful error.

Resolution order, highest priority first:

1. Object's own file_base always wins

```
[Outputs]
  file_base = common_name
  [a]
    type = Exodus
    file_base = a_own_name
  []
[]
→ a_own_name.e
```

2. No file_base anywhere — default naming (unchanged)
```
[Outputs]
  exodus = true
  [csv_a]
    type = CSV
  []
[]
→ <input>_out.e, <input>_csv_a.csv
```
3. Common file_base set, object doesn't set its own
3a. Shortcut syntax → always verbatim
```
[Outputs]
  file_base = input
  exodus = true
[]
→ input.e
```
3b. Sub-block, no collision → verbatim
```
[Outputs]
  file_base = input
  [csv]
    type = CSV
  []
[]
→ input.csv
```

3c. Sub-block, collision → error
```
[Outputs]
  file_base = input
  exodus = true
  [exodus_a]
    type = Exodus
  []
[]
→
The output object 'exodus_a' would write to the file 'input.e', which is also used by
another output object, because both fall back to the common 'file_base' set in the
[Outputs] block. MOOSE will not silently rename an output to resolve this collision.
Set 'append_object_name = true' on 'exodus_a' (or on the other colliding output(s)) to
have the object name appended and disambiguate the filenames, or give 'exodus_a' its
own distinct 'file_base'.
```

Resolving 3c — option A: append_object_name = true
```
[Outputs]
  file_base = input
  exodus = true
  [exodus_a]
    type = Exodus
    append_object_name = true
  []
[]
→ input.e, input_exodus_a.e
```

Resolving 3c — option B: distinct file_base
```
[Outputs]
  file_base = input
  exodus = true
  [exodus_a]
    type = Exodus
    file_base = a_distinct_name
  []
[]
→ input.e, a_distinct_name.e
```

---

Out of scope: identical output names (not file_base) still error unconditionally via the pre-existing uniqueness check, since appending an object name can't disambiguate two objects that already share the same name:
```
[Outputs]
  csv = true
  [csv]
    type = CSV
  []
[]
→ mooseError (name collision on csv, unrelated to file_base)
```

Summary:

| Case | Configuration | Result |
|---|---|---|
| Own `file_base` wins | `[a]` sets `file_base = a_own_name` itself | `a_own_name.e` |
| No `file_base` anywhere (default naming, unchanged) | `exodus = true` shortcut; `[csv_a] type = CSV` sub-block | `<input>_out.e`, `<input>_csv_a.csv` |
| Common `file_base`, shortcut syntax | `file_base = input` + `exodus = true` | `input.e` (always verbatim, never appended) |
| Common `file_base`, sub-block, no collision | `file_base = input` + `[csv] type = CSV` | `input.csv` (verbatim, untouched) |
| Common `file_base`, sub-block collides with another fallback output | `file_base = input` + `exodus = true` + `[exodus_a] type = Exodus` | Errors — won't silently rename; requires `append_object_name = true` or a distinct `file_base` |
| Collision resolved via `append_object_name` | Same as above, with `append_object_name = true` on `exodus_a` | `input.e`, `input_exodus_a.e` |
| Collision resolved via distinct `file_base` | Same as above, with `file_base = a_distinct_name` on `exodus_a` | `input.e`, `a_distinct_name.e` |
| Out of scope: identical output names (not `file_base`) | `csv = true` + `[csv] type = CSV` | Errors unconditionally (pre-existing `hasOutput` check; unrelated to this fix) |

**Comment by kyriv-lab** [comment]
@GiudGiud should we close this issue?

**Comment by GiudGiud** [comment]
not until it's the default behavior. 
Currently this is opt-in
